---
title: "RNASeq analysis"
output: html_document
date: "2025-02-04"
---

Libraries to load
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(DESeq2)
library(tidyverse)
library(apeglm)
library(pheatmap)
library(RColorBrewer)

```

Loading in the data
```{r readdata}
setwd("/Users/mckenzir/Desktop/Others/RNAseq/")
sample_info <- read.csv('sample_info.csv')

load("~/Desktop/Others/RNAseq/counts_matrix.RData")
load("~/Desktop/Others/RNAseq/filtered_counts_matrix.RData")

```

Changing row names, plus checking row and column names
```{r DESeq, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

rownames(sample_info) <- c('1b_1h_IL12','1h_1h_IL12','1b_1h_IL18','1h_6h_null','1b_6h_IL18','1e_6h_IL18',
                           '1e_1h_IL18','1h_6h_IL18','1b_6h_IL12','1e_6h_null','1h_1h_IL18','1b_6h_null',
                           '1e_1h_IL12','1h_6h_IL12','1e_1h_null','1b_1h_null','1h_1h_aCD3','1e_6h_aCD3',
                           '1h_6h_aCD3','1h_1h_null','1e_1h_aCD3','1b_1h_aCD3','1e_6h_IL12','1b_6h_aCD3')

#Or use: rownames(sample_info) <- as.character(sample_info$Sample_name)

all(colnames(counts_matrix) %in% rownames(sample_info))
all(colnames(counts_matrix) == rownames(sample_info))
identical(colnames(counts_matrix), rownames(sample_info))
```

Design set up for DESeq
Mouse corrects for both sex and the day of the experiment whilst Time_treatment accounts for both the time and treatment

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts_matrix, 
                              colData = sample_info,
                              design = ~ Mouse + Time_treatment)

```

Pre-filtering to keep genes that have over 200 counts and are present in at least 3 samples

```{r}
smallestGroupSize <- 3
keep <- rowSums(counts(dds) >= 200) >= smallestGroupSize 
sum(keep) # 9738 genes to keep
sum(!keep) # 16114 genes to remove
ddskeep <- dds[keep,]
```

Setting the reference level
```{r eval=FALSE}
ddskeep$Time_treatment <- factor(dds$Time_treatment, levels = c("1h_CD3", "1h_PBS"))

#ddskeep$condition <- relevel(ddskeep$condition, ref = "1h_PBS")

#"6h_PBS", "1h_CD3", "6h_CD3", "1h_IL12", "6h_IL12", "1h_IL18", "6h_IL18"

#droplevel to check if samples are blank
ddskeep$Time_treatment <- droplevels(dds$Time_treatment)

ddskept <- DESeq(ddskeep)
ddskept
res <- results(ddskept)
res

res <- results(ddskept, contrast=c("Time_treatment","1h_CD3","1h_PBS"))

resultsNames(ddskept)
```

```{r}
ddskeep$Time_treatment <- relevel(ddskeep$Time_treatment, ref = "1h_PBS")
ddskept <- DESeq(ddskeep) #ANDREW- run all on everything, we can then pull the contrasts we want later. 
resultsNames(ddskept) #ANDREW- We can see all the contrasts that have been tested and can grab whatever we want, interesting that not every possibility is here, I'll have to check how DESEQ numerators and denominators work
# Arianne - It should all still work. The "reference" is 1h PBS in the first mouse now, so if you want to compare to that, you can just grab the coefficient or you can name the contrast. To compare any other set, name the contrast as you're doing.

#ANDREW- can set contrasts as before to get specific comparisons
res_1h_PBS_vs_1h_CD3 <- results(ddskept, contrast=c("Time_treatment", "1h_CD3", "1h_PBS"))
res_1h_PBS_vs_1h_CD3 <- res_1h_PBS_vs_1h_CD3[order(res_1h_PBS_vs_1h_CD3$padj),]

# It's worth writing out the results to tables for future investigation and making supplemental tables. Example here:
write.csv(res_1h_PBS_vs_1h_CD3, file='DE_1h_PBSvCD3.csv', quote=FALSE)

#ANDREW- Or use results 'name' to achieve the same thing
res_1h_PBS_vs_1h_CD3 <- results(ddskept, name="Time_treatment_1h_CD3_vs_1h_PBS") 
res_1h_PBS_vs_1h_CD3[order(res_1h_PBS_vs_1h_CD3$padj),]

#ANDREW- Can get any contrast we want from the same object
res_1h_IL18_vs_1h_CD3 <- results(ddskept, contrast=c("Time_treatment", "1h_IL18", "1h_CD3"))
res_1h_IL18_vs_1h_CD3[order(res_1h_IL18_vs_1h_CD3$padj),]

#ANDREW- We can check if Mouse is driving any differential expression which we need to worry about
#res_mousebh <- results(ddskept, contrast=c("Mouse", "ARAA15.1b", "ARAA15.1h"))
#res_mousebh[order(res_mouse$padj),]
#"ARAA15.1e"

#ANDREW- Or if sex is confounding our other differentials
#res_Male_1h_CD3 <- results(ddskept, contrast=list(c("Time_treatment_1h_PBS_vs_1h_CD3", "SexMale.Time_treatment1h_PBS")))
#res_Male_1h_CD3[order(res_Male_1h_CD3$padj),]

#ANDREW- is effect of treatment different in males instead of females? 
#res_Male_1h_IL12 <- results(ddskept, name="SexMale.Time_treatment1h_IL12")
#res_Male_1h_IL12[order(res_Male_1h_IL12$padj),]

res <- res_1h_PBS_vs_1h_CD3 #ANDREW- So later parts of the script still function
```

<!-- ANDREW- I'm not actually acquantainted with LFC and apeglm, so we'll see if I can also learn something here
Arianne - basically this makes the lfcs smaller for visualisation purposes. Use the stats above as the actual statistics, but no issue running the plotting below with shurken lfcs. -->
#LFC
```{r}
resLFC <- lfcShrink(ddskept, coef="Time_treatment_1h_CD3_vs_1h_PBS", type="apeglm") #ANDREW-I changed the coefficient since I changed the design earlier
resLFC
head(resLFC[order(resLFC$padj),])
summary(resLFC)

plotMA(resLFC)
idx <- identify(resLFC$baseMean, resLFC$log2FoldChange) #ANDREW- I'm honestly not sure what this achieves...Arianne - gives ability to click on specific dots on the plot
rownames(resLFC)[idx]
```


```{r}
ddsLRT <- DESeq(ddskept, test="LRT", reduced= ~ Mouse) 
#resLRT <- results(ddsLRT, name ="Time_treatment_1h_PBS_vs_1h_CD3") #ANDREW-Added contrast
resLRT <- results(ddsLRT)
resLRT
summary(resLRT)

plotMA(resLRT)
idx <- identify(resLRT$baseMean, resLRT$log2FoldChange)
rownames(resLRT)[idx]
```

#MAplot for these results
plotMA(res05, ylim=c(-2,2))

plotMA(res05, ylim=c(-10,3))

idx <- identify(res05$baseMean, res05$log2FoldChange)
rownames(res05)[idx]

#8 = Ifng, 7 = Tnf



plotCounts(ddskept, gene=which.min(res$padj), intgroup="Time_treatment")

plotCounts(ddskept, "Ifng",intgroup="Time_treatment")
plotCounts(ddskept, "Tnf",intgroup="Time_treatment")
plotCounts(ddskept, "Gzmb",intgroup="Time_treatment")

#A nicer plot
library("ggplot2")
d <- plotCounts(ddskept, "Ifng",intgroup="Time_treatment", 
                returnData=TRUE)
ggplot(d, aes(x=Time_treatment, y=count)) + 
  geom_point(position=position_jitter(w=0.1,h=0)) +
  scale_y_log10(breaks=c(0,1000,5000,10000,50000,100000,500000))

e <- plotCounts(ddskept, "Tnf",intgroup="Time_treatment", 
                returnData=TRUE)
ggplot(e, aes(x=Time_treatment, y=count)) + 
  geom_point(position=position_jitter(w=0.1,h=0)) +
  scale_y_log10(breaks=c(0,1000,5000,10000,50000))
  
  f <- plotCounts(ddskept, "Gzmb",intgroup="Time_treatment", 
                returnData=TRUE)
ggplot(f, aes(x=Time_treatment, y=count)) + 
  geom_point(position=position_jitter(w=0.1,h=0)) +
  scale_y_log10(breaks=c(0,1000,10000,100000,1000000,10000000,100000000,1000000000,10000000000))
  

```{r Comparisons between conditions, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#6h PBS vs 1h PBS
resPBS <- results(ddskept, alpha=0.05, contrast=c("Time_treatment","6h_PBS","1h_PBS"))
resPBS
summary(resPBS)
resPBS <- resPBS[order(res$pvalue),]

#adj p-value<0.05, LFC 1447 up, 1470 down

plotMA(resPBS, ylim=c(-7,3))

idx <- identify(resPBS$baseMean, resPBS$log2FoldChange)
rownames(resPBS)[idx]


#1h CD3 vs 1h PBS
res1hPBSCD3 <- results(ddskept, alpha=0.05, contrast=c("Time_treatment","1h_CD3","1h_PBS"))
res1hPBSCD3
summary(res1hPBSCD3)
res1hPBSCD3 <- res1hPBSCD3[order(res$pvalue),]

#adj p-value<0.05, LFC 55 up, 0 down

plotMA(res1hPBSCD3, ylim=c(-2,11))

idx <- identify(res1hPBSCD3$baseMean, res1hPBSCD3$log2FoldChange)
rownames(res1hPBSCD3)[idx]


#1h IL12 vs 1h PBS
res1hPBSIL12 <- results(ddskept, alpha=0.05, contrast=c("Time_treatment","1h_IL12","1h_PBS"))
res1hPBSIL12
summary(res1hPBSIL12)
res1hPBSIL12 <- res1hPBSIL12[order(res$pvalue),]

#adj p-value<0.05, LFC 26 up, 1 down

plotMA(res1hPBSIL12, ylim=c(-3,5))

idx <- identify(res1hPBSIL12$baseMean, res1hPBSIL12$log2FoldChange)
rownames(res1hPBSIL12)[idx]


#1h IL18 vs 1h PBS
res1hPBSIL18 <- results(ddskept, alpha=0.05, contrast=c("Time_treatment","1h_IL18","1h_PBS"))
res1hPBSIL18
summary(res1hPBSIL18)
res1hPBSIL18 <- res1hPBSIL18[order(res$pvalue),]

#adj p-value<0.05, LFC 0 up, 0 down

plotMA(res1hPBSIL18, ylim=c(-2,2))

idx <- identify(res1hPBSIL18$baseMean, res1hPBSIL18$log2FoldChange)
rownames(res1hPBSIL18)[idx]


#6h CD3 vs 6h PBS
res6hPBSCD3 <- results(ddskept, alpha=0.05, contrast=c("Time_treatment","6h_CD3","6h_PBS"))
res6hPBSCD3
summary(res6hPBSCD3)
res6hPBSCD3 <- res6hPBSCD3[order(res$pvalue),]

#adj p-value<0.05, LFC 3006 up, 2680 down

plotMA(res6hPBSCD3, ylim=c(-3,15))

idx <- identify(res6hPBSCD3$baseMean, res6hPBSCD3$log2FoldChange)
rownames(res6hPBSCD3)[idx]


#6h IL12 vs 6h PBS
res6hPBSIL12 <- results(ddskept, alpha=0.05, contrast=c("Time_treatment","6h_IL12","6h_PBS"))
res6hPBSIL12
summary(res6hPBSIL12)
res6hPBSIL12 <- res6hPBSIL12[order(res$pvalue),]

#adj p-value<0.05, LFC 684 up, 517 down

plotMA(res6hPBSIL12, ylim=c(-4,8))

idx <- identify(res6hPBSIL12$baseMean, res6hPBSIL12$log2FoldChange)
rownames(res6hPBSIL12)[idx]


#6h IL18 vs 6h PBS
res6hPBSIL18 <- results(ddskept, alpha=0.05, contrast=c("Time_treatment","6h_IL18","6h_PBS"))
res6hPBSIL18
summary(res6hPBSIL18)
res6hPBSIL18 <- res6hPBSIL18[order(res$pvalue),]

#adj p-value<0.05, LFC 4 up, 0 down

plotMA(res6hPBSIL18, ylim=c(-2,6))

idx <- identify(res6hPBSIL18$baseMean, res6hPBSIL18$log2FoldChange)
rownames(res6hPBSIL18)[idx]


#6h CD3 vs 1h CD3
res6hcd3CD3 <- results(ddskept, alpha=0.05, contrast=c("Time_treatment","6h_CD3","1h_CD3"))
res6hcd3CD3
summary(res6hcd3CD3)
res6hcd3CD3 <- res6hcd3CD3[order(res$pvalue),]

#adj p-value<0.05, LFC 3069 up, 3103 down

plotMA(res6hcd3CD3, ylim=c(-5,12))

idx <- identify(res6hcd3CD3$baseMean, res6hcd3CD3$log2FoldChange)
rownames(res6hcd3CD3)[idx]


#6h IL12 vs 1h IL12
res6hil12IL12 <- results(ddskept, alpha=0.05, contrast=c("Time_treatment","6h_IL12","1h_IL12"))
res6hil12IL12
summary(res6hil12IL12)
res6hil12IL12 <- res6hil12IL12[order(res$pvalue),]

#adj p-value<0.05, LFC 1881 up, 1792 down

plotMA(res6hil12IL12, ylim=c(-7,6))

idx <- identify(res6hil12IL12$baseMean, res6hil12IL12$log2FoldChange)
rownames(res6hil12IL12)[idx]


#6h IL18 vs 1h IL18
res6hil18IL18 <- results(ddskept, alpha=0.05, contrast=c("Time_treatment","6h_IL18","1h_IL18"))
res6hil18IL18
summary(res6hil18IL18)
res6hil18IL18 <- res6hil18IL18[order(res$pvalue),]

#adj p-value<0.05, LFC 1493 up, 1447 down

plotMA(res6hil18IL18, ylim=c(-7,5))

idx <- identify(res6hil18IL18$baseMean, res6hil18IL18$log2FoldChange)
rownames(res6hil18IL18)[idx]

  
```{r Heatmap DESeq, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

install.packages('pheatmap')
library("pheatmap")

ntd <- normTransform(ddskept)
#Need normalised data for the heatmap
#Can do these transformations too
vsd <- vst(ddskept, blind=FALSE)
rld <- rlog(ddskept, blind=FALSE)

select <- order(rowMeans(counts(ddskept,normalized=TRUE)),
                decreasing=TRUE)[1:20]
df <- as.data.frame(colData(ddskept)[,c("Time_treatment","Mouse")])
pheatmap(assay(ntd)[select,], cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=FALSE, annotation_col=df)

#Can do these heatmaps with the other data transformations too
pheatmap(assay(vsd)[select,], cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=FALSE, annotation_col=df)

pheatmap(assay(rld)[select,], cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=FALSE, annotation_col=df)
         
#Sample to sample distances
sampleDists <- dist(t(assay(vsd)))

install.packages("RColorBrewer")
library("RColorBrewer")

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$Sample_name, sep="-")
colnames(sampleDistMatrix) <- paste(vsd$Sample_name, sep="-")
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

#PCA plot
plotPCA(vsd, intgroup=c("Time", "Treatment"))

plotPCA(vsd, intgroup=c("Time_treatment", "Replicate"))

plotPCA(vsd, intgroup=c("Sex", "Mouse"))

#Time and treatment
pcaData <- plotPCA(vsd, intgroup=c("Time", "Treatment"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=Time, shape=Treatment)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()

#Sex and Mouse
pcaData2 <- plotPCA(vsd, intgroup=c("Sex", "Mouse"), returnData=TRUE)
percentVar2 <- round(100 * attr(pcaData2, "percentVar"))
ggplot(pcaData2, aes(PC1, PC2, color=Sex, shape=Mouse)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar2[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar2[2],"% variance")) + 
  coord_fixed()

```{r Contrasts DESeq, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#All seems like a repeat of above
#In the MAplot section

results(ddskept, contrast=c("Time_treatment","1h_PBS" ,"1h_CD3" ))
#contrast = c(factor, numerator, denominator)

resultsNames(ddskept)
#"Intercept""Time_treatment_1h_IL12_vs_1h_CD3""Time_treatment_1h_IL18_vs_1h_CD3""Time_treatment_1h_PBS_vs_1h_CD3""Time_treatment_6h_CD3_vs_1h_CD3""Time_treatment_6h_IL12_vs_1h_CD3""Time_treatment_6h_IL18_vs_1h_CD3" "Time_treatment_6h_PBS_vs_1h_CD3" 

#Interactions

ddskept$Time_treatment <- factor(paste0(ddskept$Time, ddskept$Treatment))
design(ddskept) <- ~ Time_treatment
ddsK <- DESeq(ddskept)
resultsNames(ddsK)

#Potentially have to do all "1hPBS", "6hPBS", "1hCD3", "6hCD3", "1hIL12", "6hIL12", "1hIL18", "6hIL18"
results(ddsK, contrast=c("Time_treatment", "1hPBS", "1hCD3"))

PCa <- plotCounts(ddsK, 'Ifng', intgroup="Time_treatment", 
                returnData=TRUE)
                
ggplot(PCa, aes(x=Time_treatment, y=count)) + 
  geom_point(position=position_jitter(w=0.1,h=0)) + scale_y_log10(breaks=c(0, 1000,5000,10000,50000,100000))



```{r multifactor DESeq, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
#Don't think this works for the data set
#ends up being the same as the analysis above

colData(dds)
ddsMF <- dds

levels(ddsMF$Time_treatment)

levels(ddsMF$Time_treatment) <- sub("_-.*", "", levels(ddsMF$Time_treatment))
levels(ddsMF$Time_treatment)

design(ddsMF) <- formula(~ Time_treatment)
ddsMF <- DESeq(ddsMF)

resMF <- results(ddsMF)
head(resMF)

```{r Other DESeq, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
#Need to run this again to check it all and compare it to the analysis above

dds <- DESeqDataSetFromMatrix(countData = counts_matrix,
                              colData = sample_info,
                              design = ~ Time_treatment)

ddsDE <- DESeq(dds)

resultsNames(ddsDE)

keepdds_200 <- rowSums(counts(dds)) >= 200

ddskeep <- dds[keepdds_200,]

dds$Time_treatment <- factor(dds$Time_treatment, levels = c("PBS"))

ddsDET_T <- DESeq(dds2)
resT_T <- results(ddsDET_T)

summary(resT_T)

resT_T0.05 <- results(ddsDET_T, alpha = 0.05)
summary(resT_T0.05)


```{r DESeq with Time and Treatment, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
#Need to run this again to check it all and compare it to the analysis above

dds2 <- DESeqDataSetFromMatrix(countData = counts_matrix,
                              colData = sample_info,
                              design = ~ Time + Treatment)

#Pre-filtering

smallestGroupSize <- 4
keep2 <- rowSums(counts(dds2) >= 200) >= smallestGroupSize
dds2keep <- dds2[keep2,]

#Reference level

resultsNames(dds2kept)

dds2keep$Time <- factor(dds2$Time, levels = c("1h"))
#"6h_PBS", "1h_CD3", "6h_CD3", "1h_IL12", "6h_IL12", "1h_IL18", "6h_IL18"

#droplevel to check if samples are blank
dds2keep$Time_treatment <- droplevels(dds2$Time_treatment)

dds2kept <- DESeq(dds2keep)
dds2kept
res2 <- results(dds2kept)
res2

resultsNames(dds2kept)

#significance of 0.05
res205 <- results(dds2kept, alpha=0.05)
summary(res205)
res205
res205 <- res205[order(res$pvalue),]


#MAplot for these results
plotMA(res205, ylim=c(-2,2))

plotMA(res205, ylim=c(-15,2))

idx <- identify(res205$baseMean, res205$log2FoldChange)
rownames(res205)[idx]

vsd2 <- vst(dds2kept, blind=FALSE)


plotPCA(vsd2, intgroup=c("Time", "Treatment"))

plotPCA(vsd2, intgroup=c("Sex", "Mouse"))




ddsDE2 <- DESeq(dds2)
resultsNames(ddsDE2)

keepdds2_200 <- rowSums(counts(dds2)) >= 200
dds2keep <- dds2[keepdds2_200,]

dds2$Time_Treatment <- factor(dds$Time_treatment, levels = c("PBS"))

#Filtered to remove any genes with fewer than 200 counts

dds2keep$Time_Treatment <- factor(dds$Time_treatment, levels = c("PBS"))
dds2keepDET_T <- DESeq(dds2keep)
res2keepT_T0.05 <- results(dds2keepDET_T, alpha = 0.05)
summary(res2keepT_T0.05)

```{r cars}

```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}

```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
